src/main/groovy/com/github/winplay02/gitcraft/manifest/skyrising/SkyrisingMetadataProvider.java

```diff
@@ -0,0 +1,109 @@
+	}
+
+	@Override
+	public String getInternalName() {
+		return "skyrising";
+	}
+
+	@Override
+	protected OrderedVersion loadVersionFromManifest(SkyrisingManifest.VersionEntry manifestEntry, Path targetDir) throws IOException {
+		VersionInfo info = this.fetchVersionMetadata(manifestEntry.id(), manifestEntry.url(), manifestEntry.sha1(), targetDir.resolve("info"), "version info", VersionInfo.class);
+		VersionDetails details = this.fetchVersionMetadata(manifestEntry.id(), manifestEntry.details(), null, targetDir.resolve("details"), "version details", VersionDetails.class);
+		this.versionDetails.put(details.id(), details);
+		return OrderedVersion.from(info, details.normalizedVersion());
+	}
+
+	@Override
+	protected void loadVersionsFromRepository(Path dir, Consumer<OrderedVersion> loader) throws IOException {
+		Set<VersionInfo> infos = new LinkedHashSet<>();
+		Map<String, VersionDetails> detailses = new LinkedHashMap<>();
+
+		for (Path file : Files.newDirectoryStream(dir, f -> Files.isRegularFile(f) && f.endsWith(".json"))) {
+			VersionInfo info = this.loadVersionMetadata(file, VersionInfo.class);
+
+			// we could check every field but this ought to be enough
+			if (info.id() != null && info.assets() != null) {
+				infos.add(info);
+			} else {
+				VersionDetails details = this.loadVersionMetadata(file, VersionDetails.class);
+
+				// we could check every field but this ought to be enough
+				if (details.id() != null && details.normalizedVersion() != null) {
+					detailses.put(details.id(), details);
+				}
+			}
+		}
+
+		for (VersionInfo info : infos) {
+			VersionDetails details = detailses.remove(info.id());
+
+			if (details != null) {
+				this.versionDetails.put(details.id(), details);
+				loader.accept(OrderedVersion.from(info, details.normalizedVersion()));
+			} else {
+				MiscHelper.println("repository contains version info json for %s, but no version details json!", info.id());
+			}
+		}
+		for (VersionDetails details : detailses.values()) {
+			MiscHelper.println("repository contains version details json for %s, but no version info json!", details.id());
+		}
+	}
+
+	@Override
+	protected boolean isExistingVersionMetadataValid(SkyrisingManifest.VersionEntry manifestEntry, Path targetDir) throws IOException {
+		return this.isExistingVersionMetadataValid(manifestEntry.id(), manifestEntry.url(), manifestEntry.sha1(), targetDir.resolve("info"))
+			&& this.isExistingVersionMetadataValid(manifestEntry.id(), manifestEntry.details(), null, targetDir.resolve("details"));
+	}
+
+	@Override
+	public List<String> getParentVersion(OrderedVersion mcVersion) {
+		return this.getVersionDetails(mcVersion.launcherFriendlyVersionName()).previous().stream()
+			.filter(versionId -> switch (versionId) {
+				// 12w34a has 1.3.2 as parent while 12w32a has 1.3.1 as parent
+				// leading to two valid paths through those versions
+				case "1.3.2"     -> false;
```

Comment:
Is that really an issue? If I see it correctly, it will just branch and merge again.

```
1.3.1 --> 12w32a ----> 12w34a
    \---> 1.3.2 -------/
```

Reply:
It's a symptom of a different issue, which is that there is nothing in place for picking between multiple paths of the same length. It picks one based on the order in which parent versions are read. In this case it would pick the `1.3.1 -> 1.3.2 -> 12w34a` path and leave 12w32a on a side branch. It's related to the issue above with marking the main branch, there should be some logic to picking paths if multiple valid options exist.